src/mame/f256/f256.cpp

```diff
@@ -0,0 +1,2008 @@
+//             {
+//                 m_timer0_val = 0;
+//             }
+
+//             if (m_timer0_val == cmp)
+//             {
+//                 m_timer0_eq = 1;
+//                 logerror("TIMER0 up value reached %X\n", cmp);
+//             }
+
+//         }
+//         else
+//         {
+//             // down
+//             m_timer0_val--;
+//             // roll over to 24 bits
+//             if (m_timer0_val == 0xFFFF'FFFF)
+//             {
+//                 m_timer0_val = 0xFF'FFFF;
+//             }
+//             if (m_timer0_val == cmp)
+//             {
+//                 m_timer0_eq = 1;
+//                 logerror("TIMER0 down value reached %X\n", cmp);
+//             }
+//         }
+//         if (m_timer0_eq == 1 && (reg_t0 & 0x80) !=0)
+//         {
+//             timer0_interrupt_handler(1);
+//         }
+//     }
+// }
+
+
+// This timer is much slower than Timer0, so we can use single increments
+TIMER_CALLBACK_MEMBER(f256_state::timer1)
+{
+    uint8_t reg_t1 = m_iopage0->read(0xD658 - 0xC000);
+    uint32_t cmp = m_iopage0->read(0xD65D - 0xC000) + (m_iopage0->read(0xD65E - 0xC000) << 8) +
+            (m_iopage0->read(0xD65F - 0xC000) << 16);
+    logerror("TIMER1 event %06X CMP: %06X\n", m_timer1_val, cmp);
+    // if timer as reached value, then execute the action
+    if (m_timer1_eq == 1)
+    {
+        int8_t action = m_iopage0->read(0xD65C - 0xC000);
+        if (action & 1)
+        {
+            logerror("TIMER1 Cleared\n");
+            m_timer1_val = 0;
+        }
+        else
+        {
+            m_timer1_val = m_timer1_load;
+            logerror("TIMER1 Reloaded with %X\n", m_timer1_val);
+        }
+        m_timer1_eq = 0;
+    }
+    else
+    {
+        if ((reg_t1 & 8) != 0)
+        {
+            // up
+            m_timer1_val++;
+
+            // it's a 24 bit register
+            if (m_timer1_val == 0x100'0000)
+            {
+                m_timer1_val = 0;
+            }
+
+            if (m_timer1_val == cmp)
+            {
+                m_timer1_eq = 1;
+                logerror("TIMER1 up value reached %X\n", cmp);
+            }
+
+        }
+        else
+        {
+            // down
+            m_timer1_val--;
+            // roll over to 24 bits
+            if (m_timer1_val == 0xFFFF'FFFF)
+            {
+                m_timer1_val = 0xFF'FFFF;
+            }
+            if (m_timer1_val == cmp)
+            {
+                m_timer1_eq = 1;
+                logerror("TIMER1 up value reached %X\n", cmp);
+            }
+        }
+        if (m_timer1_eq == 1 && (reg_t1 & 0x80) !=0)
+        {
+            timer1_interrupt_handler(1);
+        }
+    }
+}
+
+//-------------------------------------------------
+//  VIA0 - JOYSTICK
+//-------------------------------------------------
+u8 f256_state::via0_system_porta_r()
+{
+    //logerror("VIA #0 Port A Read ioport JOY2: %02X\n", data);
+    return ioport("JOY2")->read();
```

Comment:
These port reads should be using `required_ioport` or similar finder objects. (Actually, these should ultimately be refactored to use slot devices, which would make the full range of console-compatible controllers available.)